src/combinations_up_to.rs

```diff
@@ -0,0 +1,89 @@
+use crate::{Combinations, Itertools};
+use std::fmt;
+use std::iter::FusedIterator;
+
+/// An iterator to iterate through all the up to `k`-length combinations in an iterator.
+///
+/// See [`.combinations_up_to()`](Itertools::combinations_up_to) for more information.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct CombinationsUpTo<I: Iterator> {
+    iter: Combinations<I>,
+    orig_iter: I,
+    end: bool,
+}
+
+impl<I> Clone for CombinationsUpTo<I>
+    where
+        I: Clone + Iterator,
+        I::Item: Clone,
+{
+    clone_fields!(iter, orig_iter, end);
+}
+
+impl<I> fmt::Debug for CombinationsUpTo<I>
+    where
+        I: fmt::Debug + Iterator,
+        I::Item: fmt::Debug,
+{
+    debug_fmt_fields!(CombinationsUpTo, iter, orig_iter, end);
+}
+
+#[inline]
+pub fn combinations_up_to<I>(iter: I, k: usize) -> CombinationsUpTo<I>
+    where
+        I: Iterator + Sized + Clone,
+        I::Item: Clone,
+{
+    CombinationsUpTo {
+        iter: iter.clone().combinations(k),
+        orig_iter: iter,
+        end: false,
+    }
+}
+
+impl<I> Iterator for CombinationsUpTo<I>
+    where
+        I: Iterator + Clone,
+        I::Item: Clone,
+{
+    type Item = Vec<I::Item>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.end {
+            None
+        } else if let Some(result) = self.iter.next() {
+            if result.is_empty() {
+                self.end = true;
+            }
```

Comment:
Do you think we can get rid of the `end` field by checking if `self.iter.k() == 0`?